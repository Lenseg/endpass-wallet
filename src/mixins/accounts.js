// Mixin for common methods for Ethereum accounts

export default {
  computed: {
    activeAccount() {
      return this.$store.state.accounts.activeAccount;
    },
    address() {
      return this.activeAccount.getAddressString();
    },
    balance() {
      return this.$store.state.accounts.balance === null ? null : web3.utils.fromWei(this.$store.state.accounts.balance);
    }
  }
}
