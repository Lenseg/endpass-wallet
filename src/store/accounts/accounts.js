export default {
  namespaced: true,
  state: {
    accounts: [],
    activeAccount: null
  },
  mutations: {
    add(state, account) {
      state.accounts.push(account);
      state.activeAccount = state.accounts[state.accounts.length - 1];
    },
    remove(state, index) {
      if(state.activeAccount === state.accounts[index]) {
        state.activeAccount = state.accounts.length ? state.accounts[0] : null;
      }
      state.accounts.splice(index,1);
    },
    select(state, index) {
      state.activeAccount = state.accounts[index];
    }
  }
}
