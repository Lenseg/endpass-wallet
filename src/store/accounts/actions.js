import { userService } from '@/services';
import web3 from '@/utils/web3';
import Bip39 from 'bip39';
import HDKey from 'ethereumjs-wallet/hdkey';
import { hdKeyMnemonic, kdfParams } from '@/config';
import EthWallet from 'ethereumjs-wallet';
import { Wallet } from '@/class';
import keystore from '@/utils/keystore';
import {
  SAVE_TOKENS,
  SAVE_TRACKED_TOKENS,
} from '@/store/tokens/mutations-types';
import {
  SET_ADDRESS,
  SET_WALLET,
  ADD_WALLET,
  SET_HD_KEY,
  SET_BALANCE,
  SET_SETTINGS,
  SET_OTP_SETTINGS,
  SET_EMAIL,
  ADD_ADDRESS,
} from './mutations-types';

const { toChecksumAddress } = web3.utils;

const selectWallet = async ({ commit, state, dispatch }, address) => {
  commit(SET_WALLET, state.wallets[address]);
  commit(SET_ADDRESS, address);
  dispatch('updateBalance');

  return dispatch('tokens/subscribeOnTokensBalancesUpdates', null, {
    root: true,
  });
};

const addWallet = async ({ commit, dispatch }, json) => {
  try {
    const address = web3.utils.toChecksumAddress(json.address);
    const updatedJSON = { ...json, address };

    await userService.setAccount(address, updatedJSON);
    commit(ADD_WALLET, updatedJSON);
  } catch (e) {
    dispatch('errors/emitError', e, { root: true });
  }
};

const addWalletAndSelect = async ({ dispatch }, json) => {
  try {
    await dispatch('addWallet', json);
    await dispatch('selectWallet', toChecksumAddress(json.address));
  } catch (e) {
    dispatch('errors/emitError', e, { root: true });
  }
};

// Import wallet from json V3 keystore
const addWalletWithV3 = async ({ dispatch }, { json, password }) => {
  try {
    const wallet = new Wallet(json);
    const privateKey = await wallet.getPrivateKeyString(password);

    return dispatch('addWalletWithPrivateKey', { privateKey, password });
  } catch (e) {
    return dispatch('errors/emitError', e, { root: true });
  }
};

const addWalletWithPrivateKey = async (
  { dispatch },
  { privateKey, password },
) => {
  try {
    const wallet = EthWallet.fromPrivateKey(
      Buffer.from(privateKey.replace(/^0x/, ''), 'hex'),
    );
    const json = keystore.encryptWallet(password, wallet);

    return dispatch('addWalletAndSelect', json);
  } catch (e) {
    return dispatch('errors/emitError', e, { root: true });
  }
};

const addWalletWithPublicKey = async (
  { commit, dispatch },
  publicKeyOrAddress,
) => {
  // TODO convert public key to address, accept xPub key
  try {
    const address = web3.utils.toChecksumAddress(publicKeyOrAddress);
    await userService.setAccount(address, null);
    commit(ADD_ADDRESS, address);

    return dispatch('selectWallet', address);
  } catch (e) {
    return dispatch('errors/emitError', e, { root: true });
  }
};

const generateWallet = async ({ dispatch, state, getters }, password) => {
  if (!state.hdKey) {
    throw new Error('hdKey doesn`t exist');
  }

  const hdWallet = getters.hdWallet(password);
  const i = Object.keys(state.wallets).length;
  const wallet = hdWallet.deriveChild(i).getWallet();
  const json = keystore.encryptWallet(password, wallet);

  await dispatch('addWalletAndSelect', json);
};

// Saves HD wallet's extended keys on the server
const saveHdWallet = (ctx, json) => userService.setAccount(json.address, json);

const addHdWallet = async ({ commit, dispatch }, { key, password }) => {
  try {
    const seed = Bip39.mnemonicToSeed(key);
    const hdKey = HDKey.fromMasterSeed(seed);
    const hdWallet = hdKey.derivePath(hdKeyMnemonic.path);
    // Encrypt extended private key
    const json = keystore.encryptHDWallet(password, hdWallet);
    commit(SET_HD_KEY, json);

    // Save HD keys and generate the first child wallet
    await dispatch('saveHdWallet', json);
    await dispatch('generateWallet', password);
  } catch (e) {
    dispatch('errors/emitError', e, { root: true });
  }
};

const addMultiHdWallet = async ({ dispatch }, { key, password }) => {
  const seed = Bip39.mnemonicToSeed(key);
  const hdKey = HDKey.fromMasterSeed(seed);
  const hdWallet = hdKey.derivePath(hdKeyMnemonic.path);

  /* eslint-disable no-await-in-loop */
  for (let index = 0; index < 5; index++) {
    const wallet = hdWallet.deriveChild(index).getWallet();
    const walletV3 = wallet.toV3(Buffer.from(password), kdfParams);
    const { address } = walletV3;

    if (index === 0) {
      dispatch('addWalletAndSelect', walletV3);
    } else {
      dispatch('addWallet', walletV3);
    }

    try {
      const balance = await web3.eth.getBalance(address);

      if (balance === '0') {
        break;
      }
    } catch (e) {
      break;
    }
  }
  /* eslint-enable no-await-in-loop */
};

const updateBalance = async ({ commit, dispatch, state }) => {
  if (!state.address) {
    return;
  }

  try {
    const address = state.address.getChecksumAddressString();
    const balance = await web3.eth.getBalance(address);
    commit(SET_BALANCE, balance);
  } catch (e) {
    dispatch('errors/emitError', e, { root: true });
  }
};

const updateSettings = async ({ commit, dispatch }, settings) => {
  try {
    commit(SET_SETTINGS, settings);
    await userService.setSetting('settings', settings);
  } catch (e) {
    dispatch('errors/emitError', e, { root: true });
  }
};

const validatePassword = ({ state }, password) =>
  state.wallet.validatePassword(password);

const login = (ctx, email) => userService.login(email);

const logout = async ({ commit, dispatch }) => {
  try {
    commit(SET_EMAIL, null);
    await userService.logout();
    window.location.reload();
  } catch (e) {
    dispatch('errors/emitError', e, { root: true });
  }
};

const loginViaOTP = (ctx, { code, email }) =>
  userService.loginViaOTP(code, email);

const getOtpSettings = async ({ commit, dispatch }) => {
  try {
    const otpSettings = await userService.getOtpSettings();
    commit(SET_OTP_SETTINGS, otpSettings);
  } catch (e) {
    dispatch('errors/emitError', e, { root: true });
  }
};

const setOtpSettings = async ({ commit, dispatch }, { secret, code }) => {
  try {
    await userService.setOtpSettings(secret, code);
    commit(SET_OTP_SETTINGS, { status: 'enabled' });
  } catch (e) {
    dispatch('errors/emitError', e, { root: true });
  }
};

const deleteOtpSettings = async ({ commit, dispatch }, { code }) => {
  try {
    await userService.deleteOtpSettings(code);
    commit(SET_OTP_SETTINGS, {});
    await dispatch('getOtpSettings');
  } catch (e) {
    dispatch('errors/emitError', e, { root: true });
  }
};

const setUserSettings = async ({ commit, dispatch }) => {
  try {
    const { settings, email, tokens } = await userService.getSettings();

    if (email) {
      commit(SET_EMAIL, email);
    }

    if (settings) {
      commit(SET_SETTINGS, settings);
    }

    if (tokens) {
      commit(`tokens/${SAVE_TOKENS}`, tokens, {
        root: true,
      });
      // Saved token contract addresses on all networks
      const tokenAddrs = []
        .concat(...Object.values(tokens))
        .map(token => token.address);
      commit(`tokens/${SAVE_TRACKED_TOKENS}`, tokenAddrs, { root: true });
    }
  } catch (e) {
    await dispatch('errors/emitError', e, { root: true });
  }
};

const setUserHdKey = async ({ commit, dispatch }) => {
  try {
    // Fetch and save HD wallet
    const hdKey = await userService.getHDKey();

    if (hdKey) {
      commit(SET_HD_KEY, hdKey);
    }
  } catch (e) {
    await dispatch('errors/emitError', e, { root: true });
  }
};

const setUserWallets = async ({ commit, dispatch }) => {
  try {
    // Fetch and save regular accounts
    const accounts = await userService.getV3Accounts();

    if (accounts && accounts.length) {
      accounts.forEach(account => {
        if (keystore.isV3(account)) {
          // Encrypted private key
          commit(ADD_WALLET, account);
        } else {
          // Read-only public key
          commit(ADD_ADDRESS, account.address);
        }
      });
      await dispatch('selectWallet', accounts[0].address);
    }
  } catch (e) {
    await dispatch('errors/emitError', e, { root: true });
  }
};

const init = async ({ dispatch }) => {
  try {
    await Promise.all([
      dispatch('setUserSettings'),
      dispatch('setUserHdKey'),
      dispatch('setUserWallets'),
    ]);
  } catch (e) {
    await dispatch('errors/emitError', e, { root: true });
  }
};

export default {
  selectWallet,
  addWallet,
  addWalletAndSelect,
  addWalletWithV3,
  addWalletWithPrivateKey,
  addWalletWithPublicKey,
  generateWallet,
  setUserHdKey,
  setUserWallets,
  saveHdWallet,
  addHdWallet,
  addMultiHdWallet,
  updateBalance,
  validatePassword,

  updateSettings,
  login,
  logout,
  loginViaOTP,
  getOtpSettings,
  setOtpSettings,
  setUserSettings,
  deleteOtpSettings,

  init,
};
