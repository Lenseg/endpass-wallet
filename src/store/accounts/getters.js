import web3 from 'web3';
import { BigNumber } from 'bignumber.js';
import keystore from '@/utils/keystore';
import { Wallet, Address } from '@/class';

const getAccountAddresses = state =>
  Object.keys(state.wallets).map(wallet => wallet.toLowerCase());

const isPublicAccount = state =>
  state.address instanceof Address && !(state.wallet instanceof Wallet);

// TODO move to user store?
const isLoggedIn = state => !!state.email;

const balance = (state, getters, rootState, rootGetters) => {
  if (state.balance === null) return null;

  const pendingBalance = rootGetters['transactions/pendingBalance'];
  const balanceWei = BigNumber(state.balance || '0')
    .minus(pendingBalance)
    .toFixed();

  return web3.utils.fromWei(balanceWei);
};

// Returns a decrypted HD Wallet
const hdWallet = state => password => {
  if (!state.hdKey) {
    return null;
  }

  return keystore.decryptHDWallet(password, state.hdKey);
};

export default {
  getAccountAddresses,
  isPublicAccount,
  isLoggedIn,
  balance,
  hdWallet,
};
