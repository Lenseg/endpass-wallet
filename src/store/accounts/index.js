import actions from './actions';
import getters from './getters';
import mutations from './mutations';

const state = {
  // TODO move to user store?
  email: null,
  // The encrypted xprv key for the user's hd wallet as a V3 keystore
  // object. All accounts are direct child accounts.
  hdKey: null,
  // The user's wallets, keyed by address
  wallets: {},
  // The currently selected wallet
  wallet: null,
  address: null,
  balance: null,
  // prettier-ignore
  availableCurrencies: ['USD', 'AUD', 'BRL', 'CAD', 'CHF', 'CLP', 'CNY', 'CZK', 'DKK', 'EUR', 'GBP', 'HKD', 'HUF', 'IDR', 'ILS', 'INR', 'JPY', 'KRW', 'MXN', 'MYR', 'NOK', 'NZD', 'PHP', 'PKR', 'PLN', 'RUB', 'SEK', 'SGD', 'THB', 'TRY', 'TWD', 'ZAR'],
  settings: {
    fiatCurrency: 'USD',
  },
  otpSettings: {
    secret: null,
    status: null,
  },
};

export default {
  namespaced: true,
  state,
  actions,
  getters,
  mutations,
};
