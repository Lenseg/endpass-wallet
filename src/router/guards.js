import store from '../store';
import storage from '@/services/storage';

export function privateWalletGuard(to, from, next) {
  if (store.state.accounts.wallet) {
    next();
  } else {
    next(from.fullPath);
  }
}

export function hasLoginGuard(to, from, next) {
  const { user } = store.state;
  const redirectUri = {
    path: '/',
    query: {
      redirect_uri: to.fullPath,
    },
  };

  if (user.authorizationStatus === null) {
    storage
      .readAll()
      .then(() => next(user.authorizationStatus ? undefined : redirectUri));
  } else {
    next(user.authorizationStatus ? undefined : redirectUri);
  }
}
