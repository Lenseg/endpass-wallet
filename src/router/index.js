import Vue from 'vue'
import Router from 'vue-router'
import multiguard from 'vue-router-multiguard';

import HomePage from '@/components/pages/Home'
import SendPage from '@/components/pages/Send'
import HistoryPage from '@/components/pages/History'
import ReceivePage from '@/components/pages/Receive'
import TokensPage from '@/components/pages/Tokens'
import NewWallet from '@/components/pages/NewWallet'
import ImportWallet from '@/components/pages/ImportWallet'
import ExportWallet from '@/components/pages/ExportWallet'

import store from '../store'
Vue.use(Router)

export default new Router({
  routes: [
    {
      path: '/',
      name: 'HomePage',
      component: HomePage
    },
    {
      path: '/send',
      name: 'SendPage',
      component: SendPage,
      beforeEnter: multiguard([hasWalletGuard, privateWalletGuard])
    },
    {
      path: '/tokens',
      name: 'TokensPage',
      component: TokensPage,
      beforeEnter: hasWalletGuard
    },
    {
      path: '/history',
      name: 'HistoryPage',
      component: HistoryPage,
      beforeEnter: hasWalletGuard
    },
    {
      path: '/receive',
      name: 'ReceivePage',
      component: ReceivePage,
      beforeEnter: hasWalletGuard
    },
    {
      path: '/new',
      name: 'NewWallet',
      component: NewWallet,
      beforeEnter: noWalletGuard
    },
    {
      path: '/import',
      name: 'ImportWallet',
      component: ImportWallet,
      beforeEnter: noWalletGuard
    },
    {
      path: '/export',
      name: 'ExportWallet',
      component: ExportWallet,
      beforeEnter: multiguard([hasWalletGuard, privateWalletGuard])
    }
  ]
})
function hasWalletGuard (to, from, next) {
  if(store.state.accounts.activeAccount) {
    next();
  } else {
    next(from.fullPath);
  }
}

function noWalletGuard(to, from, next) {
  if(!store.state.accounts.activeAccount) {
    next();
  } else {
    next(from.fullPath);
  }
}

function privateWalletGuard(to, from, next) {
  if(!store.state.accounts.activeAccount._privKey !== null) {
    next();
  } else {
    next(from.fullPath);
  }
}
