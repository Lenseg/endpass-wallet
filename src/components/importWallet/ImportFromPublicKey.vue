<template>
  <v-form :submitHandler="addWalletWithAddress">
    <v-input v-model="address"
             label="Address"
             id="address"
             name="address"
             validator="required|address"
             data-vv-as="private key"
             key="publicKeyUnique"
             aria-describedby="address"
             placeholder="0x...."
             required />

    <v-button className="is-primary is-cta"
              :loading="isCreating">Import</v-button>
  </v-form>
</template>

<script>
import router from '@/router';
import { mapActions, mapMutations } from 'vuex';
import VForm from '@/components/ui/form/VForm.vue';
import VInput from '@/components/ui/form/VInput.vue';
import VButton from '@/components/ui/form/VButton.vue';

export default {
  name: 'import-from-public-key',
  data: () => ({
    isCreating: false,
    address: '',
  }),
  methods: {
    ...mapMutations('accounts', ['addAddress']),
    async addWalletWithAddress() {
      this.isCreating = true;
      await new Promise(res => setTimeout(res, 20));
      try {
        this.addAddress(this.address);
        router.push('/');
      } catch (e) {
        this.errors.add({
          field: 'address',
          msg: 'Address is invalid',
          id: 'wrongAddress',
        });
        console.error(e);
      }
      this.isCreating = false;
    },
  },
  components: {
    VForm,
    VInput,
    VButton,
  },
};
</script>

<style lang="scss">
</style>
