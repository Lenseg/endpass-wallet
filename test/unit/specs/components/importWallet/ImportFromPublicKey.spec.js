import { shallow, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';
import VeeValidate from 'vee-validate';

import { address } from '../../../fixtures/accounts';

import ImportFromPublicKey from '@/components/importWallet/ImportFromPublicKey';

import router from '@/router';

jest.mock('@/router', () => require('../../../__mocks__/router'));

const localVue = createLocalVue();

localVue.use(Vuex);
localVue.use(VeeValidate);

describe('ImportFromPublicKey', () => {
  let wrapper, actions;
  beforeEach(() => {
    actions = {
      addWalletWithPublicKey: jest.fn(),
    };
    const storeOptions = {
      modules: {
        accounts: {
          namespaced: true,
          actions,
        },
      },
    };
    const store = new Vuex.Store(storeOptions);
    wrapper = shallow(ImportFromPublicKey, {
      localVue,
      store,
    });
  });
  describe('render', () => {
    it('should be a Vue component', () => {
      expect(wrapper.isVueInstance()).toBeTruthy();
    });
  });

  describe('logic', () => {
    it('should call vuex addWalletWithPublicKey with correct arguments', () => {
      wrapper.setData({ address });
      wrapper.vm.addWallet();
      expect(actions.addWalletWithPublicKey.mock.calls[0][1]).toEqual(address);
    });

    it('should redirect to root after successful wallet creation', () => {
      wrapper.vm.addWallet();
      expect(router.push).toHaveBeenCalledWith('/');
    });

    it('should add error to field if failed to create wallet', () => {
      actions.addWalletWithPublicKey.mockImplementationOnce(() => {
        throw new Error();
      });
      wrapper.vm.addWallet();
      expect(wrapper.vm.errors.items[0]).toEqual({
        field: 'address',
        id: 'wrongAddress',
        msg: 'Address is invalid',
        // vee validate added field
        scope: null,
      });
    });
  });
});
