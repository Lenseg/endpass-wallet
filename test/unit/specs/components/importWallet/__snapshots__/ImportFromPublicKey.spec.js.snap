// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImportFromPublicKey render should match snapshot 1`] = `<!---->`;
